packages/components/src/edit-in-place-control/index.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useRef, useEffect } from '@wordpress/element';
+import { ENTER, ESCAPE } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import Button from '../button';
+
+function EditInPlaceControl( { label = '', onClick = noop, onChange = noop } ) {
+	const [ edit, setEdit ] = useState( false );
+	const [ value, setValue ] = useState( label );
+	const prevValue = useRef();
+
+	const inputRef = useRef();
+	const buttonRef = useRef();
+
+	useEffect( () => {
+		prevValue.current = value;
+		if ( edit ) {
+			inputRef.current.focus();
+			inputRef.current.select();
+		} else {
+			buttonRef.current.focus();
+		}
+	}, [ edit ] );
+
+	return (
+		<>
+			{ ! edit && (
+				<Button
+					ref={ buttonRef }
+					className="components-edit-in-place-control__label"
+					onClick={ () => {
+						setEdit( true );
+						onClick();
+					} }
+				>
+					{ value }
+				</Button>
+			) }
+			{ edit && (
+				<input
+					ref={ inputRef }
+					className="components-edit-in-place-control__input"
+					value={ value }
+					onChange={ ( event ) => {
+						setValue( event.target.value );
+					} }
+					onBlur={ () => {
+						setEdit( false );
+						onChange( value );
+					} }
+					onKeyDown={ ( event ) => {
+						if (
+							ENTER === event.keyCode ||
+							ESCAPE === event.keyCode
+						) {
+							event.preventDefault();
+							event.stopPropagation();
+							if ( ESCAPE === event.keyCode ) {
+								setValue( prevValue.current );
```

Comment:
Having to wrangle `prevValue` is a little cumbersome. Instead of having this ref, we could display `label` in the button when `edit` is `false` and call `setValue( label )` when `edit` transitions to `true`.